File: solutions/elixir/raindrops/1/lib/raindrops.ex
defmodule Raindrops do
  @sounds %{
    3 => "Pling",
    5 => "Plang",
    7 => "Plong"
  }
  @doc """
  Returns a string based on raindrop factors.

  - If the number contains 3 as a prime factor, output 'Pling'.
  - If the number contains 5 as a prime factor, output 'Plang'.
  - If the number contains 7 as a prime factor, output 'Plong'.
  - If the number does not contain 3, 5, or 7 as a prime factor,
    just pass the number's digits straight through.
  """
  @spec convert(pos_integer) :: String.t()
  def convert(number) do
    sounds = @sounds
    |> Enum.map(fn({factor, sound}) ->
      if rem(number, factor) == 0, do: sound, else: ""
    end)
    |> Enum.join("")
    case sounds do
      "" -> Integer.to_string(number)
      _ -> sounds
    end
  end
end

File: solutions/erlang/raindrops/1/src/raindrops.erl
-module(raindrops).

-export([convert/1]).

convert(Number) when Number rem 3 * Number rem 5 * Number rem 7 /= 0 -> integer_to_list(Number);
convert(Number) -> 
    SomePrimes = [3, 5, 7],
    Divisors = lists:filter(fun(N) -> Number rem N == 0 end, SomePrimes),
    Sounds = lists:map(fun sound/1, Divisors),
    lists:concat(Sounds).

sound(N) when N == 3 -> "Pling";
sound(N) when N == 5 -> "Plang";
sound(N) when N == 7 -> "Plong".
